platforms/database.go

```diff
@@ -0,0 +1,77 @@
+package platforms
+
+import (
+	"runtime"
+	"strings"
+)
+
+// These function are generated from from https://golang.org/src/go/build/syslist.go.
+//
+// We use switch statements because they are slightly faster than map lookups
+// and use a little less memory.
+
+// isKnownOS returns true if we know about the operating system.
+//
+// The OS value should be normalized before calling this function.
+func isKnownOS(os string) bool {
+	switch os {
+	case "android", "darwin", "dragonfly", "freebsd", "linux", "nacl", "netbsd", "openbsd", "plan9", "solaris", "windows", "zos":
+		return true
+	}
+	return false
+}
+
+// isKnownArch returns true if we know about the arcitecture.
+//
+// The Arch value should be normalized before being passed to this function.
+func isKnownArch(arch string) bool {
+	switch arch {
+	case "386", "amd64", "amd64p32", "arm", "armbe", "arm64", "arm64be", "ppc64", "ppc64le", "mips", "mipsle", "mips64", "mips64le", "mips64p32", "mips64p32le", "ppc", "s390", "s390x", "sparc", "sparc64":
+		return true
+	}
+	return false
+}
+
+func normalizeOS(os string) string {
+	if os == "" {
+		return runtime.GOOS
+	}
+	os = strings.ToLower(os)
+
+	switch os {
+	case "macos":
+		os = "darwin"
+	}
+	return os
+}
+
+// normalizeArch normalizes the architecture.
+func normalizeArch(arch, variant string) (string, string) {
+	arch = strings.ToLower(arch)
+	switch arch {
+	case "i386":
+		arch = "386"
+		variant = ""
+	case "x86_64", "x86-64":
+		arch = "amd64"
+		variant = ""
+	case "aarch64":
+		arch = "arm64"
+		variant = "" // v8 is implied
+	case "armhf":
+		arch = "arm"
+		variant = ""
```

Comment:
We normalise (below) `linux/arm` into `linux/arm/v7` but here we are normalizing `linux/armhf` into just `linux/arm`, which seems a bit non-normalysed to me.

IOW if the result of `normalizeArch` was actually normative (as I think it should be) then the assertion in the following should always be true:
```
foo := normalizeArch(input)
assert(foo == normalizeArch(foo))
```
but that doesn't appear to be the case when `input == "linux/armhf"`

Reply:
TL;DR: I think you can ignore my previous comment.

I think I misread the code handling the "arm" case, that switch doesn't actually have the `default` case I thought it did so `normalizeArch("arm") == "arm"` as required.

(I'm also aware I was taking some liberties with the actual arguments, there should be `split("/", foo)` stuff sprinkled around in my examples to make them closer to actually working examples)

Reply:
Just added a test case to ensure the round trip.